### PR TITLE
[BPF] fix ct cleaning for host processes w/o ctlb using service

### PR DESCRIPTION
**Pull request by tomastigera**
If there is an overlay, forwarding entry may not have a matching client IP as it may start with host's main IP, but then it is turned into host's tunnel IP. We cannot determine what is the client/service/backend from that entry only. We need to lookup the reverse entry. If the reverse entry does not exist, we delete the forarding entry anyway as there is no use for it without the reverse half.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf/conntrack/cleanup.go [hunk 1]
```diff
@@ -288,8 +291,16 @@ func (sns *StaleNATScanner) Check(k KeyInterface, v ValueInterface, _ EntryGet)
...
+				rv, err := get(revKey)
+				if err != nil {
+					// There is no match for the reverse key, delete it, its useless - we
+					// can get here due to host networked program accessing service
+					// without ctlb when the backed is accessible via tunnel.
+					return ScanVerdictDelete
+				}
+				k = revKey
+				v = rv
+				goto reverse // handle it based on the reverse entry
```
Comment (fasaxc) [inline felix/bpf/conntrack/cleanup.go:303]: Think I'd prefer recursion to a goto (at least that would die with a stack overflow if there was an infinite loop).

File: felix/bpf/conntrack/cleanup.go [hunk 2]
```diff
@@ -288,8 +291,16 @@ func (sns *StaleNATScanner) Check(k KeyInterface, v ValueInterface, _ EntryGet)
 				svcIP = kA
 				svcPort = kAport
 			} else {
-				log.WithFields(log.Fields{"key": k, "value": v}).Error("Mismatch between key and rev key")
-				return ScanVerdictOK // don't touch, will get deleted when expired
+				rv, err := get(revKey)
+				if err != nil {
```
Comment (fasaxc) [inline felix/bpf/conntrack/cleanup.go:295]: Can we check that it was a "not exists" error?

File: felix/bpf/conntrack/cleanup.go [hunk 3]
```diff
@@ -288,8 +291,16 @@ func (sns *StaleNATScanner) Check(k KeyInterface, v ValueInterface, _ EntryGet)
...
+					// There is no match for the reverse key, delete it, its useless - we
+					// can get here due to host networked program accessing service
+					// without ctlb when the backed is accessible via tunnel.
+					return ScanVerdictDelete
```
Comment (fasaxc) [inline felix/bpf/conntrack/cleanup.go:299]: Suggest at least a debug log here; if we ever have a bug in cleanup, good to have a log for every deletion so we can trace it back

**Review by fasaxc — CHANGES_REQUESTED** [review]
Couple of suggestions.

How does the BPF program behave with these mismatched entries?  Is it truly broken in that case?  could we avoid _creating_ a broken entry in if we know that'll be the case

File: felix/bpf/conntrack/cleanup.go (hunk 1 above)
Comment (tomastigera) [reply felix/bpf/conntrack/cleanup.go:303]: `reverse` checks what the type is to avoid infinite looping. Will never jump back to this.

**Review by fasaxc — APPROVED** [review]
Don't think I need to re-review now you explained the fix.

**Pull request merged**